### ISSUE TITLE
Iterator by default do not sort within bucket?

### ISSUE THREAD

**Issue by fred2008** [pytorch/text#536]
According to the example code in the ducument, the bucketIterator cannot sort according to sort_key at all. The sort_key is not used when running and therefore cannot achieve "Minimizes amount of padding needed while producing freshly shuffled batches for each new epoch. See pool for the bucketing procedure used."  as [https://torchtext.readthedocs.io/en/latest/data.html#BucketIterator](url).
``` python
# continuing from above
mt_dev = data.TranslationDataset(
    path='data/mt/newstest2014', exts=('.en', '.de'),
    fields=(src, trg))
src.build_vocab(mt_train, max_size=80000)
trg.build_vocab(mt_train, max_size=40000)
# mt_dev shares the fields, so it shares their vocab objects

train_iter = data.BucketIterator(
    dataset=mt_train, batch_size=32,
    sort_key=lambda x: data.interleave_keys(len(x.src), len(x.trg)))
# usage
>>>next(iter(train_iter))
<data.Batch(batch_size=32, src=[LongTensor (32, 25)], trg=[LongTensor (32, 28)])>
```

should the example be as follows to Sort within buckets and reduce the pad number?
```python
train_iter = data.BucketIterator(
    dataset=mt_train, batch_size=32,
    sort_key=lambda x: data.interleave_keys(len(x.src), len(x.trg)),
    sort_within_batch=True
)
```

**Comment by gordicaleksa** [comment]
Yep, that's true. Without sort_within_batch enabled nothing really happens. It just chunks examples into 100*batch_size and then continues to take batch_size batches without any sorting whatsoever!

You could also set the "sort" to True which will initially sort the examples according to the sort_key function.

I had to dig into the source code to know how to use it - and that's definitely not a good library design.

**Comment by zhangguanheng66** [comment]
Thanks for the feedback. Just FYI, we are going to retire `Iterator` and `Batch` entirely soon. User should use `DataLoader` to generate batch (`collate_fn`) and iterate.

**Comment by abhinavarora** [comment]
Closing issue as we have gotten rid of iterators and the issue is no more relevant. 
#1537
